### ISSUE TITLE
RT-DETR Validation hyperparameters have no effect on val metrics

### ISSUE THREAD

**Issue by DavyBee** [ultralytics/ultralytics#20128]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Val

### Bug

When validating with a fine-tuned RTDETR model, the hyperparameters (such as conf in this example) have no effect on model metrics output by model.val(). They seem to work when predicting with the model, but so far have had no effect on any metrics values. Attached below are examples with my own trained model, and also the retrained model, executing model.val() on ‘coco.yaml’. The one with conf = 0.2 produces the same metrics as the model with conf = 0.99. The outputs of the model on coco are not all 0.99, so I know that this output does not make sense. 

My model:

<img width="1332" alt="Image" src="https://github.com/user-attachments/assets/c432caca-f400-4492-9d36-1a0b5bf77574" />

Pretrained model too: 

<img width="1251" alt="Image" src="https://github.com/user-attachments/assets/d5997b93-563b-4908-bb2e-8bc35111f3f3" />

<img width="1242" alt="Image" src="https://github.com/user-attachments/assets/1ed9129a-6544-4392-a033-da8665d4509f" />

### Environment

Ultralytics 8.3.107 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
Setup complete ✅ (2 CPUs, 12.7 GB RAM, 41.3/235.7 GB disk)

OS                  Linux-6.1.85+-x86_64-with-glibc2.35
Environment         Colab
Python              3.11.12
Install             pip
Path                /usr/local/lib/python3.11/dist-packages/ultralytics
RAM                 12.67 GB
Disk                41.3/235.7 GB
CPU                 Intel Xeon 2.00GHz
CPU count           2
GPU                 Tesla T4, 15095MiB
GPU count           1
CUDA                12.4

numpy               ✅ 2.0.2<=2.1.1,>=1.23.0
matplotlib          ✅ 3.10.0>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.1.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.6.0+cu124>=1.8.0
torch               ✅ 2.6.0+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.21.0+cu124>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 5.9.5
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0
{'OS': 'Linux-6.1.85+-x86_64-with-glibc2.35',
 'Environment': 'Colab',
 'Python': '3.11.12',
 'Install': 'pip',
 'Path': '/usr/local/lib/python3.11/dist-packages/ultralytics',
 'RAM': '12.67 GB',
 'Disk': '41.3/235.7 GB',
 'CPU': 'Intel Xeon 2.00GHz',
 'CPU count': 2,
 'GPU': 'Tesla T4, 15095MiB',
 'GPU count': 1,
 'CUDA': '12.4',
 'Package Info': {'numpy': '✅ 2.0.2<=2.1.1,>=1.23.0',
  'matplotlib': '✅ 3.10.0>=3.3.0',
  'opencv-python': '✅ 4.11.0.86>=4.6.0',
  'pillow': '✅ 11.1.0>=7.1.2',
  'pyyaml': '✅ 6.0.2>=5.3.1',
  'requests': '✅ 2.32.3>=2.23.0',
  'scipy': '✅ 1.14.1>=1.4.1',
  'torch': '✅ 2.6.0+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"',
  'torchvision': '✅ 0.21.0+cu124>=0.9.0',
  'tqdm': '✅ 4.67.1>=4.64.0',
  'psutil': '✅ 5.9.5',
  'py-cpuinfo': '✅ 9.0.0',
  'pandas': '✅ 2.2.2>=1.1.4',
  'seaborn': '✅ 0.13.2>=0.11.0',
  'ultralytics-thop': '✅ 2.0.14>=2.0.0'}}

### Minimal Reproducible Example

model = RTDETR('rtdetr-l.pt')

metrics1 = model.val(data = 'coco.yaml', conf = 0.2)
metrics2 = model.val(data = 'coco.yaml', conf = .99)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @DavyBee, thank you for bringing this to our attention! 🚀 Your detailed report is much appreciated.

We recommend reviewing the [Docs](https://docs.ultralytics.com/) for guidance on validation processes and hyperparameter usage. For example, you can explore the [Python](https://docs.ultralytics.com/usage/python/) or [CLI](https://docs.ultralytics.com/usage/cli/) sections for helpful examples. If you're working with custom training, you might also find the [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) valuable.

If this is indeed a 🐛 Bug Report, please ensure you’ve included a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to allow us to replicate and debug the issue effectively. From your post, it seems you’ve provided a great starting point—thank you! If there's any additional context or observations you can share, please do.

For broader discussions or help from the community, feel free to join us on:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time chat
- [Discourse](https://community.ultralytics.com/) for in-depth discussions
- [Reddit](https://reddit.com/r/Ultralytics) for community threads

## Upgrade

To ensure the issue is not already resolved, please upgrade to the latest `ultralytics` package and verify all dependencies are up-to-date. You can do so with the following command:

```bash
pip install -U ultralytics
```

## Environments

YOLO is compatible with various environments. You can try running your code in one of the following verified setups:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Please note that this is an automated response to help guide you toward self-resolving your issue. An Ultralytics engineer will review your report and provide additional assistance as soon as possible. Thank you for your patience! 😊

**Comment by Y-T-G** [comment]
Conf filter is disabled for validation 

https://github.com/ultralytics/ultralytics/blob/7becb54d59a864d9f0c11720b85e03870e3c035e/ultralytics/models/rtdetr/val.py#L146-L147

**Comment by DavyBee** [comment]
Ok yes it seems like this is the issue, but I don’t see any reason why this should be the case? Still potentially a bug? Why would you want to allow confidence filtering for prediction but not for validation? Or even more basically, why would you want to disabled confidence filtering for validation? Also, any suggestions on how to get around this when running through the ultralytics Python API?

**Comment by Y-T-G** [comment]
I don't think you're supposed to filter using confidence during validation. That would make the mAP incorrect because mAP calculation requires unfiltered detections.

**Comment by DavyBee** [comment]
Hmm, I potentially understand that, but wouldn’t you want to evaluate on the output that you would actually be working with? Isn’t the point of validation here to get metrics that I can use to fine-tune hyperparameters such as confidence for my eventual predictions on unseen data?

**Comment by Y-T-G** [comment]
For mAP, the evaluation needs the unfiltered output because that's what mAP is based on by definition. But the confusion matrix is based on a fixed threshold of 0.25.

**Comment by DavyBee** [comment]
Just coming back to this, this might make sense if it was consistent across models, however metrics DO change for YOLO models when changing these hyperparameters. I don’t follow the logic there - if it doesn’t make sense to change in one instance (RT-DETR), why can they change in another (YOLO). Pics Below

<img width="775" alt="Image" src="https://github.com/user-attachments/assets/82e34f3e-3efb-4349-b08f-ea37d42362bd" />
<img width="763" alt="Image" src="https://github.com/user-attachments/assets/9c67a95e-d6b8-430a-b37d-48bfdbf9dfe6"/>

**Comment by Y-T-G** [comment]
I opened a PR to add the confidence filter. But in either case, it breaks the mAP calculation. mAP is to be calculated on all the confidence range by definition.

**Comment by DavyBee** [comment]
Amazing, thanks for your help and explanations

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has just been merged in PR #20175. This update introduces a confidence threshold filter to the RT-DETR validator, so only predictions above the specified confidence level are considered during evaluation. This enhancement should make model assessment more accurate and consistent with real-world use. The Ultralytics package version has also been updated from 8.3.108 to 8.3.109.

Special thanks to the author and contributors for their work on this improvement!

To test if this resolves the issue, try upgrading with:
```
pip install -U ultralytics>=8.3.109
```
Alternatively, you can install the latest code from the main branch:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Or, feel free to wait for the next official PyPI release.

Please let us know if this update resolves the issue for you. Thanks for reporting it, and any further feedback is always welcome! 🙏

**Comment by glenn-jocher** [comment]
I'm glad to see this has been fixed in the latest update. To clarify a bit more on the technical side: while mAP calculations traditionally use predictions across the full confidence range, having the ability to filter by confidence during validation is valuable for evaluating model performance under real-world usage conditions.

The inconsistency between YOLO and RT-DETR validation behavior was indeed a legitimate issue, and it's great that the validation process is now more consistent across model architectures. This change will help you better tune your confidence thresholds based on validation metrics that match your deployment scenario.
